imsim/photon_ops.py

```diff
@@ -0,0 +1,140 @@
+        # )
+        self.telescope = telescope
+        self.detector = get_camera()[det_name]
+        self.boresight = boresight
+        self.sky_pos = sky_pos
+        self.image_pos = image_pos
+        self.icrf_to_field = icrf_to_field
+
+    def applyTo(self, photon_array, local_wcs=None, rng=None):
+        """Apply the photon operator to a PhotonArray.
+
+        Here, we assume that the photon array has passed through
+        `imsim.atmPSF.AtmosphericPSF` which stores sampled pupil
+        locations and abuses `dxdz`, `dydz` to store those values.
+
+        Parameters
+        ----------
+        photon_array:   A `PhotonArray` to apply the operator to.
+        local_wcs:      A `LocalWCS` instance defining the local WCS for the current photon
+                        bundle in case the operator needs this information.  [default: None]
+        rng:            A random number generator to use if needed. [default: None]
+        """
+        # x/y to u/v
+        u = local_wcs._u(photon_array.x, photon_array.y)
+        v = local_wcs._v(photon_array.x, photon_array.y)
+        # u/v to ICRF
+        u_rad = np.deg2rad(u / 3600)
+        v_rad = np.deg2rad(v / 3600)
+        ra, dec = self.sky_pos.deproject_rad(u_rad, v_rad)
+        # ICRF to field
+        thx, thy = self.icrf_to_field.radecToxy(ra, dec, units="rad")
+
+        vx, vy, vz = batoid.utils.gnomonicToDirCos(thx, thy)
+        # Adjust for refractive index of air
+        wavelength = photon_array.wavelength * 1e-9
+        n = self.telescope.inMedium.getN(wavelength)
+        vx /= n
+        vy /= n
+        vz /= n
+
+        x, y = photon_array.pupil_u, photon_array.pupil_v
+        z = self.telescope.stopSurface.surface.sag(x, y)
+        ray_vec = batoid.RayVector._directInit(
+            x,
+            y,
+            z,
+            vx,
+            vy,
+            vz,
+            t=np.zeros_like(x),
+            wavelength=wavelength,
+            flux=photon_array.flux,
+            vignetted=np.zeros_like(x, dtype=bool),
+            failed=np.zeros_like(x, dtype=bool),
+            coordSys=self.telescope.stopSurface.coordSys,
+        )
+        traced = self.telescope.trace(ray_vec)
+        ray_vector_to_photon_array(traced, detector=self.detector, out=photon_array)
+        photon_array.x -= self.image_pos.x
+        photon_array.y -= self.image_pos.y
+
+    def __str__(self):
+        return "imsim.LsstPhotonOp()"
+
+    def __repr__(self):
+        return str(self)
+
+
+def ray_vector_to_photon_array(
+    ray_vector: batoid.RayVector, detector, out: PhotonArray
+) -> PhotonArray:
+    """Converts a batoid.RayVector to a galsim.PhotonArray
+
+    Stores into an already existing galsim.PhotonArray out"""
+
+    assert all(np.abs(ray_vector.z) < 1.0e-15)
+    out.x, out.y = focal_to_pixel(ray_vector.x * 1e3, ray_vector.y * 1e3, detector)
+
+    out.dxdz = ray_vector.vx / ray_vector.vz
+    out.dydz = ray_vector.vy / ray_vector.vz
+    return out
```

Comment:
I think we need to add something like `out.flux *= (1-ray_vector.vignetted)` here.  (My fault for forgetting this before).

Reply:
Should be done now

Reply:
Is that right?  I would have thought probabilistically remove photons (or set to flux=0) rather than change their fluxes to fractional values.

Reply:
Or is ray_vector.vignetted just a 0 or 1?  If that's the case, I think this is fine.

Reply:
`ray_vector.vignetted` is a bool, so this is just zeroing the flux for geometrically obstructed paths.  Maybe there's a np.where() clause that would make this more obvious?

Reply:
`out.flux = np.where(ray_vector.vignetted, 0, out.flux)` may be clearer, but looks like it allocates instead of modifies in place.

Reply:
Maybe `out.flux[ray_vector.vignetted] = 0.`?

Reply:
But the existing one is fine too, now that I know it's a bool.

Reply:
I changed it to `out.flux[ray_vector.vignetted] = 0.` which I think is clearer.